### PR TITLE
4 | create sign up feature

### LINKED ISSUE #4: Sign up

### PR DESCRIPTION
**Pull request by kyBa-Le**

### CODE REVIEW

File: src/main/java/com/example/todolist/API/Restful/DTO/Request/RequestSignIn.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.example.todolist.API.Restful.DTO.Request;
+
+public record RequestSignIn(
```
Comment (vultpham) [inline src/main/java/com/example/todolist/API/Restful/DTO/Request/RequestSignIn.java]: Should be a noun => `SignInRequest`

File: src/main/java/com/example/todolist/API/Restful/DTO/Request/RequestSignUp.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package com.example.todolist.API.Restful.DTO.Request;
+
+import com.example.todolist.Domain.Validation.UserValidation.EmailRule;
+import com.example.todolist.Domain.Validation.UserValidation.PasswordRule;
+import com.example.todolist.Domain.Validation.UserValidation.PhoneRule;
+import com.example.todolist.Domain.Validation.UserValidation.UsernameRule;
+
+public record RequestSignUp(
```
Comment (vultpham) [inline src/main/java/com/example/todolist/API/Restful/DTO/Request/RequestSignUp.java]: Should be a noun => `SignUpRequest`

File: src/main/java/com/example/todolist/API/Restful/DTO/Response/SignUpResponse.java [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package com.example.todolist.API.Restful.DTO.Response;
+
+public record SignUpResponse(
+        String username
```
Comment (vultpham) [inline src/main/java/com/example/todolist/API/Restful/DTO/Response/SignUpResponse.java]: look like return `username` is not useful, we usually, return `id` of `user`, or just nothing, based on your frontend requirement. This time just return nothing but with 201 status code

File: src/main/java/com/example/todolist/API/Restful/HandlingException/APIError.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
+package com.example.todolist.API.Restful.HandlingException;
+
+import java.util.Date;
+
+public class APIError {
```
Comment (vultpham) [inline src/main/java/com/example/todolist/API/Restful/HandlingException/APIError.java]: the class looks good, but really, we don't need getter setter, just public all field, but make sure it `readonly` or `final`, keep it simple this time

File: src/main/java/com/example/todolist/API/Restful/HandlingException/ExistingValueException.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package com.example.todolist.API.Restful.HandlingException;
+
+import java.util.Date;
+
+public class ExistingValueException extends RuntimeException {
```
Comment (vultpham) [inline src/main/java/com/example/todolist/API/Restful/HandlingException/ExistingValueException.java]: this design is considered as smell code, we'll redesign it

File: src/main/java/com/example/todolist/API/Restful/HandlingException/GlobalException.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+@ControllerAdvice
+public class GlobalException {
```
Comment (vultpham) [inline src/main/java/com/example/todolist/API/Restful/HandlingException/GlobalException.java:11]: consider making it reusable, you will have other types of API rather than restful API

File: src/main/java/com/example/todolist/API/Restful/RestController/UserController.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+    @PostMapping("/signup")
+    public SignUpResponse signUp(@RequestBody @Valid RequestSignUp requestSignUp, HttpServletResponse response, HttpServletRequest request) {
+        User user = userService.signUp(requestSignUp.username(), requestSignUp.password(), requestSignUp.email(), requestSignUp.phone());
+        HttpSession session = request.getSession();
+        session.setAttribute("user", user);
+        return new SignUpResponse(user.getUsername());
+    }
+    @GetMapping("/signup")
+    public ResponseEntity<String> signUp(HttpServletRequest request, HttpServletResponse response) {
```
Comment (vultpham) [inline src/main/java/com/example/todolist/API/Restful/RestController/UserController.java]: this API is so unnecessary :)))

File: src/main/java/com/example/todolist/API/Restful/RestController/UserController.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+@RestController
+@RequestMapping("")
+public class UserController {
+    @Autowired
+    UserService userService;
+
+    @PostMapping("/signup")
+    public SignUpResponse signUp(@RequestBody @Valid RequestSignUp requestSignUp, HttpServletResponse response, HttpServletRequest request) {
+        User user = userService.signUp(requestSignUp.username(), requestSignUp.password(), requestSignUp.email(), requestSignUp.phone());
```
Comment (vultpham) [inline src/main/java/com/example/todolist/API/Restful/RestController/UserController.java]: use horizontal code format for better readability

File: src/main/java/com/example/todolist/API/Restful/RestController/UserController.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        HttpSession session = request.getSession();
+        session.setAttribute("user", user);
```
Comment (vultpham) [inline src/main/java/com/example/todolist/API/Restful/RestController/UserController.java]: we only need to store user id, store the whole user is unnecessary, unless you have a reason for it

File: src/main/java/com/example/todolist/API/Restful/RestController/UserController.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+import jakarta.servlet.http.HttpServletResponse;
+import jakarta.servlet.http.HttpSession;
+import jakarta.validation.Valid;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("")
+public class UserController {
```
Comment (vultpham) [inline src/main/java/com/example/todolist/API/Restful/RestController/UserController.java]: I think we should call it AuthController

File: src/main/java/com/example/todolist/API/Restful/RestController/UserController.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+        session.setAttribute("user", user);
+        return new SignUpResponse(user.getUsername());
+    }
+    @GetMapping("/signup")
+    public ResponseEntity<String> signUp(HttpServletRequest request, HttpServletResponse response) {
+        return ResponseEntity.ok("You can access to the signup page");
+    }
+    @GetMapping("/login")
+    public ResponseEntity<String> logIn(HttpServletRequest request, HttpServletResponse response) {
+        return ResponseEntity.ok("This is login page!");
```
Comment (vultpham) [inline src/main/java/com/example/todolist/API/Restful/RestController/UserController.java]: this API is so unnecessary :))), frontend will handle this, not backend, we'll implement login API in next ticket

File: src/main/java/com/example/todolist/Domain/Entity/BaseEntity.java [hunk 12]
```diff
@@ -0,0 +1,15 @@
+package com.example.todolist.Domain.Entity;
+
+import jakarta.persistence.Id;
+import jakarta.persistence.MappedSuperclass;
+import java.util.UUID;
+
+@MappedSuperclass
+public abstract class BaseEntity {
+    @Id
+    private String id = UUID.randomUUID().toString();
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Entity/BaseEntity.java]: `id` should be marked `final` => we'll need to create a protected constructor

Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Entity/BaseEntity.java]: and also `id` should be protected, if we use private, child class will not be able to access it, you will want that behavior

File: src/main/java/com/example/todolist/Domain/Entity/Task.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
+package com.example.todolist.Domain.Entity;
+
+import jakarta.persistence.*;
+
+@Entity
+@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
+public class Task extends BaseEntity {
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Entity/Task.java]: remove this. this is out of scope of this ticket

File: src/main/java/com/example/todolist/Domain/Entity/User.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+
+import java.util.List;
+
+
+@Entity
+@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
+public class User extends BaseEntity {
+    @Column(unique = true, nullable = false)
+    @UsernameRule
+    private String username;
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Entity/User.java:14]: use a new line between each field to make a better readability

File: src/main/java/com/example/todolist/Domain/Entity/User.java [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    @Column(nullable = false, unique = true)
+    @EmailRule
+    private String email;
+    @PhoneRule
+    private String phone;
+
+    @OneToMany(mappedBy = "user")
+    private List<Task> tasks;
+
+    public User() {}
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Entity/User.java]: mark this as protected not public, since this is not a business logic, every new user has to be passed to an information, we have this constructor because this is required by jakatar

File: src/main/java/com/example/todolist/Domain/Entity/User.java [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+    @PasswordRule
+    private String password;
+    @Column(nullable = false, unique = true)
+    @EmailRule
+    private String email;
+    @PhoneRule
+    private String phone;
+
+    @OneToMany(mappedBy = "user")
+    private List<Task> tasks;
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Entity/User.java]: remove this

File: src/main/java/com/example/todolist/Domain/Entity/User.java [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+        this.email = email;
+        this.phone = phone;
+    }
+
+    public String getUsername() {
+        return username;
+    }
+
+    public void setUsername(String username) {
+        this.username = username;
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Entity/User.java]: we don't allow user to update their username

File: src/main/java/com/example/todolist/Domain/Entity/User.java [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+    public void setPassword(String password) {
+        this.password = password;
+    }
+
+    public String getEmail() {
+        return email;
+    }
+
+    public void setEmail(String email) {
+        this.email = email;
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Entity/User.java]: we may not allow user to update their email address

File: src/main/java/com/example/todolist/Domain/Interceptor/AuthenticationInterceptor.java [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import jakarta.servlet.http.HttpSession;
+import org.springframework.web.servlet.HandlerInterceptor;
+import org.springframework.web.servlet.ModelAndView;
+
+import java.util.Arrays;
+import java.util.List;
+
+public class AuthenticationInterceptor implements HandlerInterceptor {
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Interceptor/AuthenticationInterceptor.java]: this should not be a part of domain

File: src/main/java/com/example/todolist/Domain/Interceptor/AuthenticationInterceptor.java [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+    private final List<String> acceptedPaths = Arrays.asList("/signup", "/login");
+
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws Exception {
+        HttpSession session = request.getSession(false);
+        String requestPath = request.getServletPath();
+        if(session != null || acceptedPaths.contains(requestPath)) {
+            return true;
+        }else{
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Interceptor/AuthenticationInterceptor.java]: bad code format

File: src/main/java/com/example/todolist/Domain/Interceptor/AuthenticationInterceptor.java [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+            response.sendRedirect("/login");
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Interceptor/AuthenticationInterceptor.java]: we don't do this, frontend will handle this, we are building API

File: src/main/java/com/example/todolist/Domain/Service/UserService.java [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.web.bind.MethodArgumentNotValidException;
+
+@Service
+public class UserService {
+    @Autowired
+    UserRepository userRepository;
+
+    public User signUp(String username, String password, String email, String phoneNumber){
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Service/UserService.java]: this should be name createUser, instead of signUp

File: src/main/java/com/example/todolist/Domain/Service/UserService.java [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+        if(userRepository.findByUsername(username) != null) {
+            throw new ExistingValueException("Username is already in use", "username");
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Service/UserService.java]: note this, I'll give feedback later

File: src/main/java/com/example/todolist/Domain/Service/UserService.java [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+        }else if(userRepository.findByEmail(email) != null) {
+            throw new ExistingValueException("Email is already in use", "email");
+        }
+        User user = new User(username, password, email, phoneNumber);
+        return userRepository.save(user);
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Service/UserService.java]: we don't save user here, just return it, controller will decide to save it

File: src/main/java/com/example/todolist/Domain/Validation/UserValidation/EmailRule.java [hunk 25]
```diff
@@ -0,0 +1,25 @@
...
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target(ElementType.FIELD)
+@Retention(RetentionPolicy.RUNTIME)
+@NotNull
+@NotBlank
+@Email(message = "Email must be written in format example@email.com")
+@Constraint(validatedBy = {})
+public @interface EmailRule {
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Validation/UserValidation/EmailRule.java:19]: note this, we will deep dive into it

File: src/main/java/com/example/todolist/Domain/Validation/UserValidation/EmailRule.java [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target(ElementType.FIELD)
+@Retention(RetentionPolicy.RUNTIME)
+@NotNull
+@NotBlank
+@Email(message = "Email must be written in format example@email.com")
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Validation/UserValidation/EmailRule.java]: note this, we will deep dive into it, 
NOTE: the message, it will not always "Email must be ..."

File: src/main/java/com/example/todolist/Domain/Validation/UserValidation/UsernameRule.java [hunk 27]
```diff
@@ -0,0 +1,23 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target(ElementType.FIELD)
+@Retention(RetentionPolicy.RUNTIME)
+@NotBlank
+@NotNull
+@Size(min=4, max = 25)
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Validation/UserValidation/UsernameRule.java:17]: TODO we will demo it, to see how it like without defining message

File: src/main/java/com/example/todolist/Domain/Validation/UserValidation/UsernameRule.java [hunk 28]
```diff
@@ -0,0 +1,23 @@
...
+@Constraint(validatedBy = {})
+public @interface UsernameRule {
+    String message() default "Invalid username";
+    Class<?>[] groups() default {};
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Validation/UserValidation/UsernameRule.java:21]: do we need this?

File: src/main/java/com/example/todolist/Domain/Validation/UserValidation/UsernameRule.java [hunk 29]
```diff
@@ -0,0 +1,23 @@
...
+import java.lang.annotation.Target;
+
+@Target(ElementType.FIELD)
+@Retention(RetentionPolicy.RUNTIME)
+@NotBlank
+@NotNull
+@Size(min=4, max = 25)
+@Constraint(validatedBy = {})
+public @interface UsernameRule {
+    String message() default "Invalid username";
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Domain/Validation/UserValidation/UsernameRule.java]: do we need this?

**Review by vultpham — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/todolist/API/Restful/RestController/UserController.java (hunk 11 above)
Comment (kyBa-Le) [reply src/main/java/com/example/todolist/API/Restful/RestController/UserController.java]: This is to test interceptor. I will delete it.

File: src/main/java/com/example/todolist/Domain/Service/UserService.java (hunk 24 above)
Comment (kyBa-Le) [reply src/main/java/com/example/todolist/Domain/Service/UserService.java]: I don't understand this idea now.

**Review by vultpham — APPROVED** [review]
LGTM
